(RE-10924) Package signing fails if already signed

mwaggett opened the pull request:
This PR includes:
- Reverting the revert of the work I did on this previously, which includes:
  - moving the rpm `sign_all` functionality out of the rake file and into the Pkg::Sign::Rpm module
  - adding a `has_sig?` check to the `sign_all` method, so we skip signing already-signed things
- Refactoring `has_sig?`, including tests (see commit message for details)

mwaggett commented:
I'm not sure what further testing we should do for this, if any. Hard to know how things will behave until they're running.. 
Also, if gpg_key is not set in some places, not sure if there's a way to track that down other than for this to go in and see what breaks.

e-gris commented on lib/packaging/sign/rpm.rb:
```diff
@@ -34,7 +34,77 @@ def legacy_sign(rpm)
+    # not set, before shelling out. We also only want the short key, all
+    # lowercase, since that's what the `rpm -Kv` output uses.
+    key = Pkg::Util::Gpg.key.downcase.chars.last(8).join
+    signature_check_output, _, _ = Pkg::Util::Execution.capture3("rpm -Kv #{rpm}")
```
Minor: use long options (when practical) when shelling out.

```rpm --checksig --verbose #{rpm}```

e-gris commented on tasks/sign.rake:
```diff
@@ -38,67 +38,8 @@ namespace :pl do
-        FileUtils.ln(path, link_path, :force => true, :verbose => true)
-      end
-    end
+    rpm_directory = args.root_dir || $DEFAULT_DIRECTORY
```
Hooray for code in rake moved into libs!

underscorgan commented on spec/lib/packaging/sign_spec.rb:
```diff
@@ -0,0 +1,106 @@
+V4 RSA/SHA256 Signature, key ID ef8d349f: NOKEY
+MD5 digest: OK (d5f06ba2a9053de532326d0659ec0d11)
+DOC
+      }
```
Can you also have sample signed responses from el5 and sles12? they have slightly different content and/or formatting, just to make sure we don't end up with anything weird.

mwaggett replied:
they look almost identical to the el7 one, but I can definitely add them

mwaggett replied:
alright, added below

underscorgan commented on spec/lib/packaging/sign_spec.rb:
```diff
@@ -0,0 +1,106 @@
+      end
+    end
+  end
+end
```
Can you also have a test for the gpg_key not being set?

underscorgan commented on lib/packaging/sign/rpm.rb:
```diff
@@ -34,7 +34,76 @@ def legacy_sign(rpm)
+    # This should allow the `Pkg::Util::Gpg.key` method to fail if gpg_key is
+    # not set, before shelling out. We also only want the short key, all
+    # lowercase, since that's what the `rpm -Kv` output uses.
+    key = Pkg::Util::Gpg.key.downcase.chars.last(8).join
```
What does this look like if gpg_key is unset? Does it throw an exception? Is the exception useful?

mwaggett replied:
yes, this gets thrown: https://github.com/puppetlabs/packaging/blob/1.0.x/lib/packaging/util/gpg.rb#L10
adding a test now

underscorgan commented:
@mwaggett as for additional testing before this can go in ... make sure you can run the signing rake tasks for a variety of projects (at least agent and some of the PE ones), and try explicitly unsetting gpg_key and make sure it does what you expect it to. Run with unsigned packages, signed packages, a group of packages where some of them are signed but not all

underscorgan commented on lib/packaging/sign/rpm.rb:
```diff
@@ -34,7 +34,76 @@ def legacy_sign(rpm)
+    # lowercase, since that's what the `rpm -Kv` output uses.
+    key = Pkg::Util::Gpg.key.downcase.chars.last(8).join
+    signature_check_output, _, _ = Pkg::Util::Execution.capture3("rpm --checksig --verbose #{rpm}")
+    return signature_check_output.include? key
```
hmm, this has the possibility that the short key will be in the md5/sha sums. can we maybe look more explicitly for "key ID #{key}"?

mwaggett replied:
yeah, I figured that would be unlikely, but that's a good call

mwaggett commented on spec/lib/packaging/sign_spec.rb:
```diff
@@ -0,0 +1,132 @@
+        allow(Pkg::Util::Execution).to receive(:capture3).and_return([unsigned_response, '', 0])
+        expect(Pkg::Sign::Rpm.has_sig?(rpm)).to be false
+      end
+      it 'fails if gpg_key is not set' do
```
@underscorgan added!

mwaggett commented on lib/packaging/sign/rpm.rb:
```diff
@@ -34,7 +34,76 @@ def legacy_sign(rpm)
+    # lowercase, since that's what the `rpm -Kv` output uses.
+    key = Pkg::Util::Gpg.key.downcase.chars.last(8).join
+    signature_check_output, _, _ = Pkg::Util::Execution.capture3("rpm --checksig --verbose #{rpm}")
+    return signature_check_output.include? "key ID #{key}"
```
@underscorgan updated!

puppetcla commented:
CLA signed by all contributors.

mwaggett commented:
**TESTING**
**from a checkout of puppet-agent 5.5.6:** 
- retrieved and signed all as usual
- cleaned out everything but rpms from the pkg dir (since this work only affects rpms and I wanted things to go faster)
- reran the sign_all task, didn't get prompted to sign again:
```
pkg/cisco-wrlinux/7/puppet5/x86_64/puppet-agent-5.5.6-1.cisco_wrlinux7.x86_64.rpm is already signed, skipping . . .
pkg/cisco-wrlinux/5/puppet5/x86_64/puppet-agent-5.5.6-1.cisco_wrlinux5.x86_64.rpm is already signed, skipping . . .
pkg/sles/11/puppet5/i386/puppet-agent-5.5.6-1.sles11.i386.rpm is already signed, skipping . . .
pkg/sles/11/puppet5/x86_64/puppet-agent-5.5.6-1.sles11.x86_64.rpm is already signed, skipping . . .
pkg/sles/12/puppet5/x86_64/puppet-agent-5.5.6-1.sles12.x86_64.rpm is already signed, skipping . . .
pkg/sles/12/puppet5/ppc64le/puppet-agent-5.5.6-1.sles12.ppc64le.rpm is already signed, skipping . . .
pkg/fedora/28/puppet5/x86_64/puppet-agent-5.5.6-1.fc28.x86_64.rpm is already signed, skipping . . .
pkg/fedora/26/puppet5/x86_64/puppet-agent-5.5.6-1.fedora26.x86_64.rpm is already signed, skipping . . .
pkg/fedora/27/puppet5/x86_64/puppet-agent-5.5.6-1.fedora27.x86_64.rpm is already signed, skipping . . .
pkg/el/6/puppet5/i386/puppet-agent-5.5.6-1.el6.i386.rpm is already signed, skipping . . .
pkg/el/6/puppet5/x86_64/puppet-agent-5.5.6-1.el6.x86_64.rpm is already signed, skipping . . .
pkg/el/7/puppet5/aarch64/puppet-agent-5.5.6-1.el7.aarch64.rpm is already signed, skipping . . .
pkg/el/7/puppet5/x86_64/puppet-agent-5.5.6-1.el7.x86_64.rpm is already signed, skipping . . .
pkg/el/7/puppet5/ppc64le/puppet-agent-5.5.6-1.el7.ppc64le.rpm is already signed, skipping . . .
pkg/el/5/puppet5/i386/puppet-agent-5.5.6-1.el5.i386.rpm is already signed, skipping . . .
pkg/el/5/puppet5/x86_64/puppet-agent-5.5.6-1.el5.x86_64.rpm is already signed, skipping . . .
```
- moved the signed rpms out of the way, retrieved again, replaced some of the newly retrieved rpms with the signed ones, reran the sign_all task, the signed ones were skipped:
```
pkg/puppet-agent-5.5.6-1.el6.i386.rpm is already signed, skipping . . .
pkg/puppet-agent-5.5.6-1.fc28.x86_64.rpm is already signed, skipping . . .
pkg/puppet-agent-5.5.6-1.el7.ppc64le.rpm is already signed, skipping . . .
pkg/puppet-agent-5.5.6-1.cisco_wrlinux5.x86_64.rpm is already signed, skipping . . .
pkg/puppet-agent-5.5.6-1.sles11.x86_64.rpm is already signed, skipping . . .
Signing legacy (v3) rpms...
Enter pass phrase:
Signing modern (v4) rpms...
Enter pass phrase:
```
- set GPG_KEY='', reran the sign_all task, it failed as expected:
```
...
rake aborted!
You need to set `gpg_key` in your build defaults.
/tmp/puppet-agent-36131231085845417/.bundle/gems/ruby/2.4.0/bundler/gems/packaging-89f6168112e7/lib/packaging/util/gpg.rb:10:in `key'
/tmp/puppet-agent-36131231085845417/.bundle/gems/ruby/2.4.0/bundler/gems/packaging-89f6168112e7/lib/packaging/sign/rpm.rb:40:in `has_sig?'
...
```
**from a checkout of pe-r10k-vanagon 2.6.2.8:**
- did the same things as for agent, behavior was the same
